### PR TITLE
retry callback

### PR DESCRIPTION
**Pull request by thallesfreitaszup**

### CODE REVIEW

File: butler/src/app/core/integrations/cd/octopipe/octopipe.service.ts [hunk 1]
```diff
@@ -29,6 +29,9 @@ import { IBaseVirtualService, IEmptyVirtualService } from '../spinnaker/connecto
 import createDestinationRules from '../spinnaker/connector/utils/manifests/base-destination-rules'
 import { createEmptyVirtualService, createVirtualService } from '../spinnaker/connector/utils/manifests/base-virtual-service'
 import { OctopipeApiService } from './octopipe-api.service'
+import {concatMap, delay, map, retryWhen, tap} from 'rxjs/operators';
```
Comment (lucasbfernandes) [inline butler/src/app/core/integrations/cd/octopipe/octopipe.service.ts:32]: Add space between brackets and names

File: butler/src/app/core/integrations/cd/octopipe/octopipe.service.ts [hunk 2]
```diff
@@ -73,6 +80,27 @@ export class OctopipeService implements ICdServiceStrategy {
...
+        return attempts >= AppConstants.CD_CONNECTION_MAXIMUM_RETRY_ATTEMPTS?
+          throwError('Reached maximum attemps.') :
+          this.getDeployRetryPipe(error, attempts)
+      })
+    )
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  private getDeployRetryPipe(error: any, attempts: number) {
+
```
Comment (lucasbfernandes) [inline butler/src/app/core/integrations/cd/octopipe/octopipe.service.ts:97]: Remove space

File: butler/src/app/core/integrations/cd/octopipe/octopipe.service.ts [hunk 3]
```diff
@@ -73,6 +80,27 @@ export class OctopipeService implements ICdServiceStrategy {
     }
   }
 
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  private getDeployRetryCondition(deployError: Observable<any>) {
+
```
Comment (lucasbfernandes) [inline butler/src/app/core/integrations/cd/octopipe/octopipe.service.ts:85]: Remove space

File: butler/src/app/core/constants/application/configuration.constants.ts [hunk 4]
```diff
@@ -36,9 +36,9 @@ export const ConfigurationConstants = {
 
   MOOVE_NOTIFICATION_MILLISECONDS_RETRY_DELAY: 1000,
 
-  SPINNAKER_CONNECTION_MAXIMUM_RETRY_ATTEMPTS : 5,
+  CD_CONNECTION_MAXIMUM_RETRY_ATTEMPTS : 5,
```
Comment (lucasbfernandes) [inline butler/src/app/core/constants/application/configuration.constants.ts:39]: Shouldn't you update spinnaker.service as well?

File: octopipe/pkg/mozart/mozart.go [hunk 5]
```diff
@@ -267,3 +267,11 @@ func (mozart *Mozart) returnPipelineError(pipelineError error) {
 		utils.CustomLog("error", "returnPipelineError", err.Error())
 	}
 }
+
+func (mozart *Mozart) SendRequest(request *http.Request, client http.Client, attempts int) (*http.Response,error) {
+	response, err := client.Do(request)
+	if err != nil && attempts< 5 {
```
Comment (lucasbfernandes) [inline octopipe/pkg/mozart/mozart.go:273]: Keep all values inside constants, do not put magic numbers inside the code (https://stackoverflow.com/questions/47882/what-is-a-magic-number-and-why-is-it-bad)

File: octopipe/pkg/mozart/mozart.go [hunk 6]
```diff
@@ -267,3 +267,11 @@ func (mozart *Mozart) returnPipelineError(pipelineError error) {
...
+		response ,err = mozart.SendRequest(request, client, attempts+1)
```
Comment (lucasbfernandes) [inline octopipe/pkg/mozart/mozart.go:274]: Put space between all operators

File: octopipe/pkg/mozart/mozart.go (hunk 5 above)
Comment (lucasbfernandes) [inline octopipe/pkg/mozart/mozart.go:273]: Put space between all operators

**Comment by thallesfreitaszup** [comment]
[Implementar retry no callback do Octopipe](https://trello.com/c/01bZqlML/527-implementar-retry-no-callback-do-octopipe)

File: octopipe/pkg/mozart/mozart.go (hunk 5 above)
Comment (thalleslmF) [reply octopipe/pkg/mozart/mozart.go]: fixed

File: octopipe/pkg/mozart/mozart.go (hunk 6 above)
Comment (thalleslmF) [reply octopipe/pkg/mozart/mozart.go]: fixed

File: octopipe/pkg/mozart/mozart.go (hunk 5 above)
Comment (thalleslmF) [reply octopipe/pkg/mozart/mozart.go]: fixed

Comment (thallesfreitaszup) [reply octopipe/pkg/mozart/mozart.go]: Fixed

File: octopipe/pkg/mozart/mozart.go (hunk 6 above)
Comment (thallesfreitaszup) [reply octopipe/pkg/mozart/mozart.go]: Fixed

File: octopipe/pkg/mozart/mozart.go (hunk 5 above)
Comment (thallesfreitaszup) [reply octopipe/pkg/mozart/mozart.go]: Fixed

**Pull request closed without merging**